libraries/AP_Camera/AP_Camera_MAVLinkCamV2.cpp

```diff
@@ -0,0 +1,215 @@
+// focus in = -1, focus hold = 0, focus out = 1
+bool AP_Camera_MAVLinkCamV2::set_manual_focus_step(int8_t focus_step)
+{
+    if (_link == nullptr) {
+        return false;
+    }
+
+    // prepare and send message
+    mavlink_command_long_t pkt {};
+    pkt.command = MAV_CMD_SET_CAMERA_FOCUS;
+    pkt.param1 = FOCUS_TYPE_CONTINUOUS; // Focus Type, 0:FOCUS_TYPE_STEP, 1:FOCUS_TYPE_CONTINUOUS, 2:FOCUS_TYPE_RANGE, 3:FOCUS_TYPE_METERS, 4:FOCUS_TYPE_AUTO, 5:FOCUS_TYPE_AUTO_SINGLE, 5:FOCUS_TYPE_AUTO_CONTINUOUS
+    pkt.param2 = focus_step;            // Focus Value
+
+    _link->send_message(MAVLINK_MSG_ID_COMMAND_LONG, (const char*)&pkt);
+
+    return true;
+}
+
+// auto focus.  returns true on success
+bool AP_Camera_MAVLinkCamV2::set_auto_focus()
+{
+    if (_link == nullptr) {
+        return false;
+    }
+
+    // prepare and send message
+    mavlink_command_long_t pkt {};
+    pkt.command = MAV_CMD_SET_CAMERA_FOCUS;
+    pkt.param1 = FOCUS_TYPE_AUTO;   // Focus Type, 0:FOCUS_TYPE_STEP, 1:FOCUS_TYPE_CONTINUOUS, 2:FOCUS_TYPE_RANGE, 3:FOCUS_TYPE_METERS, 4:FOCUS_TYPE_AUTO, 5:FOCUS_TYPE_AUTO_SINGLE, 5:FOCUS_TYPE_AUTO_CONTINUOUS
+    pkt.param2 = 0;                 // Focus Value
+
+    _link->send_message(MAVLINK_MSG_ID_COMMAND_LONG, (const char*)&pkt);
+
+    return true;
+}
+
+// handle incoming mavlink message including CAMERA_INFORMATION
+void AP_Camera_MAVLinkCamV2::handle_message(mavlink_channel_t chan, const mavlink_message_t &msg)
+{
+    // exit immediately if this is not our message
+    if (msg.sysid != _sysid || msg.compid != _compid) {
+        return;
+    }
+
+    // handle CAMERA_INFORMATION
+    if (msg.msgid == MAVLINK_MSG_ID_CAMERA_INFORMATION) {
+        mavlink_camera_information_t cam_info;
+        mavlink_msg_camera_information_decode(&msg, &cam_info);
+
+        const uint8_t fw_ver_major = cam_info.firmware_version & 0x000000FF;
+        const uint8_t fw_ver_minor = (cam_info.firmware_version & 0x0000FF00) >> 8;
+        const uint8_t fw_ver_revision = (cam_info.firmware_version & 0x00FF0000) >> 16;
+        const uint8_t fw_ver_build = (cam_info.firmware_version & 0xFF000000) >> 24;
+
+        // display camera info to user
+        gcs().send_text(MAV_SEVERITY_INFO, "Camera: %s %s fw:%u.%u.%u.%u",
+                cam_info.vendor_name,
+                cam_info.model_name,
+                (unsigned)fw_ver_major,
+                (unsigned)fw_ver_minor,
+                (unsigned)fw_ver_revision,
+                (unsigned)fw_ver_build);
+
+        _got_camera_info = true;
+    }
+}
+
+// search for camera in GCS_MAVLink routing table
+void AP_Camera_MAVLinkCamV2::find_camera()
+{
+    // do not look for camera for first 10 seconds so user may see banner
+    uint32_t now_ms = AP_HAL::millis();
+    if (now_ms < 10000) {
+        return;
+    }
+
+    // search for camera for 60 seconds or until armed
+    if ((now_ms > AP_CAMERA_MAVLINKCAMV2_SEARCH_MS) && hal.util->get_soft_armed()) {
+        return;
+    }
```

Comment:
```suggestion
    // only search for camera for first 60 seconds of boot
    if (now_ms > AP_CAMERA_MAVLINKCAMV2_SEARCH_MS) {
        return;
    }
    // don't search while armed:
    if (hal.util->get_soft_armed())) {
        return;
    }
```
This is logically "or" ,not "and as you had it.

I think that first clause is going to cause confusion in our users when they turn on their separately-power device after that  60 seconds; perhaps they have a separate avionics battery like a lot of Planes do which doesn't power their camera.

Reply:
I think what's there is consistent with what we have for the Gremsy gimbal.  I think it will search for the camera while disarmed.. and no matter what it will search for the camera for 60 seconds..

Reply:
Ah, I misunderstood the intent.  You want to search while the thing is flying but only within the first 60 seconds... useful for vehicles that are always armed...

Reply:
Yes exactly..